nan.h

```diff
@@ -1591,39 +1593,38 @@ class Callback {
 
 /* abstract */ class AsyncProgressWorker : public AsyncWorker {
  public:
+
+  typedef std::deque<std::string> async_queue_t;
+
   explicit AsyncProgressWorker(Callback *callback_)
-      : AsyncWorker(callback_), asyncdata_(NULL), asyncsize_(0) {
-    async = new uv_async_t;
+      : AsyncWorker(callback_) {
     uv_async_init(
         uv_default_loop()
-      , async
+      , &async
       , AsyncProgress_
     );
-    async->data = this;
+    async.data = this;
 
     uv_mutex_init(&async_lock);
   }
 
   virtual ~AsyncProgressWorker() {
     uv_mutex_destroy(&async_lock);
-
-    if (asyncdata_) {
-      delete[] asyncdata_;
-    }
   }
 
   void WorkProgress() {
-    uv_mutex_lock(&async_lock);
-    char *data = asyncdata_;
-    size_t size = asyncsize_;
-    asyncdata_ = NULL;
-    uv_mutex_unlock(&async_lock);
-
     // Dont send progress events after we've already completed.
     if (callback) {
-        HandleProgressCallback(data, size);
+      uv_mutex_lock(&async_lock);
+      while (!queue.empty()) {
+        std::string async_data = queue.front();
```

Comment:
I'm not quite sure, but I think we don't have to copy the string here. I think it is Ok to hold a `const&` to the string and delay the `pop_front()` until we re-aquire the lock late int the loop.

BTW, I really like this:

![screen shot 2015-11-12 at 09 24 55](https://cloud.githubusercontent.com/assets/137545/11113711/4c1bad38-891f-11e5-9e9c-9dc616a2d96b.png)

Thanks, for following through, @bfrancojr. Much appreciated.